### DISCUSSION TITLE
During the simulation of synchrotron radiation generated by an electron beam, I'm unable to output the density profile for photons.

### DISCUSSION THREAD

**Discussion by ynliu-12302020** [BLAST-WarpX/warpx#5059] [Q&A]
I am trying to calculate synchrotron radiation produced by an electron beam and further to plot the photon density profile for visualization. But the density plot always displays a value of 0(as shown in the figure below).
![photon_density_profile](https://github.com/user-attachments/assets/b6ad934f-c4c8-4a54-a9d9-f50a5d96308c)

However, it is possible to output the energy and other information of photons.
Instead of pasting the code here directly, I attached the two files. 
[input_2d.txt](https://github.com/user-attachments/files/16284271/input_2d.txt)
[photon_densityprofile_plotting.txt](https://github.com/user-attachments/files/16284272/photon_densityprofile_plotting.txt)
The 1st attachment (i.e. input_2d.txt) is the warpx input file that I am using to generate synchrotron radiation. The diagnostic output for photon density is available in a diagnostic named "diag1" in the end of the file (highlighted in red)(rho_photons).
Then, the 2nd attachment (i.e. photon_densityprofile_plotting.txt) is regarding the analysis and visualization of the photon density profile. I have successfully plotted density distribution for other particles using this script, but I am unable to plot the density distribution for photons.
Can anyone please help me pointing out the mistakes I might be doing? or help me modifying the code if necessary? I'm definitely missing something!
Many thanks!

**Comment by RemiLehe** [comment]
Hi,
Sorry about the late reply!

The issue is that the quantity that you are trying to plot (`rho_photon`) is the **charge density** of photons (not the number density). The `rho_` diagnostics is a standard diagnostic that we have in WarpX and it is meant to be applied to charged particles. It will also run without error for neutral particles, but will return zero density which admittedly is not very helpful.

Note that, in your simulations, the photons were also saved as individual macroparticles, in the same HDF5 file. When reading them as individual macroparticles, you will be able to extract their momentum (and thus energy), position, weighting factor, etc. Does that work for you? [Here](https://openpmd-viewer.readthedocs.io/en/latest/api_reference/generic_interface.html#openpmd_viewer.OpenPMDTimeSeries.get_particle) is more info on how to extract the list of macroparticles from the file.
